lib/connections/registry/registry.go

```diff
@@ -0,0 +1,66 @@
+// Copyright (C) 2019 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Registry tracks connections on which we are listening on, to allow us to pick a connection that has a NAT port
+// mapping. This also makes our outgoing port stable and same as incoming port which should allow
+// better probability of punching through.
+package registry
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/syncthing/syncthing/lib/sync"
+)
+
+var (
+	mut       = sync.NewMutex()
+	available = make(map[string][]interface{})
+)
+
+func Register(scheme string, item interface{}) {
```

Comment:
In the tcp punchthrough we need to reach into the registry from the dialer package, so this has to be generally global.

I'll do the var Default = New() thing.

Reply:
Also, the lessThan func will vary based on what's wedged into the registry.

Reply:
Ah, you mean for any given registry, with an existing set of whatevers, I might want to vary the less func at get-time? Ok then. Otherwise I think it would be neater to pass it in the constructor when the registry is created.

Reply:
Registry is literally a global map of interface{}.
Get could return a []interface{} and then it's down to the caller to sort it I guess if you prefer it that way.

Reply:
We could otherwise perhaps do a type switch in the registry and hardcode a few sorting functions.

Reply:
Riiight, so in the end each `scheme` will probably have it's own less func? I guess then it feels to me like each scheme should be a separate registry, with the less func registered there? 

```
tcpRegistry = NewRegistry("tcp", tcpLessFunc)
```

I guess of this really needs to be *global* we could have a RegistryRegistry (cue ...BeanFactoryAdapterInterface jokes).

```
registry.Global.NewSubRegistry("tcp", tcpLessFunc)
registry.Global.Add("tcp", foo)
_ = registry.Global.Get("tcp") // uses tcpLessFunc
```

I don't know if I'm making things better or worse, just thinking out loud...

Reply:
I don't really see why this is such a big deal to be honest, and why the current thing feels wrong?

Reply:
I guess we can have a QuicRegistry and a TCPRegistry etc, but that's just more globals.

Reply:
It's ok, especially if it becomes a type and not just a weird singleton package

Reply:
It's a weird singleton package, but because it needs to be, because otherwise how will dialer package reach into it?

It's ok means it's ok as it is, or it's ok and  we should split it into two globals?

Reply:
It's ok as a weird `var Default = New()` global. At least so that it's visible where the state lives. I'm tempted to say "Add a couple of tests for it" because that would make it very annoying to have it as just package magic

Reply:
It's already a var Default = New() with tests in the latest commit.